packages/react/src/List/ListItem.tsx

```diff
@@ -0,0 +1,40 @@
+import React, {type Ref, useContext} from 'react'
+import clsx from 'clsx'
+import {CheckIcon, DashIcon} from '@primer/octicons-react'
+import type {BaseProps} from '../component-helpers'
+import {ListContext} from './listContext'
+
+import styles from './List.module.css'
+
+export type ListItemProps = BaseProps<HTMLElement> & {
+  /**
+   * React.ReactNode and React.ReactNode[] are valid children.
+   */
+  children?: React.ReactNode | React.ReactNode[]
+  /**
+   * The ref object to be attached to the list item.
+   */
+  ref?: Ref<HTMLLIElement>
+}
+
+export const ListItem = ({className, children, ...props}: ListItemProps) => {
+  const {variant} = useContext(ListContext)
+
+  const leadingVisual = () => {
+    switch (variant) {
+      case 'checked':
+        return <CheckIcon className={styles['ListItem__leading-visual']} />
+      case 'default':
+        return <DashIcon className={styles['ListItem__leading-visual']} />
+
+      default:
+        return <></>
```

Comment:
This feels strange to return an empty fragment, do you want null instead to provide an actual falsey value?

Reply:
This just feels like a different approach to accomplish the same thing but maybe it is more performant? So updating in next push.

Reply:
Yeah there's probably some optimisations under the hood, but from a readability pov they're quite different, because a react fragment will return a truthy value while null be falsy.